audit/audit_test.go

```diff
@@ -0,0 +1,202 @@
+package audit
+
+import (
+	"context"
+	"github.com/ONSdigital/go-ns/identity"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/pkg/errors"
+	. "github.com/smartystreets/goconvey/convey"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+const (
+	namespace   = "audit-test"
+	auditAction = "test"
+	auditResult = "success"
+)
+
+type HandlerMock struct {
+	invocations []*http.Request
+	handleFunc  func(http.ResponseWriter, *http.Request)
+}
+
+func (h *HandlerMock) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.invocations = append(h.invocations, r)
+}
+
+func TestAuditor_Record(t *testing.T) {
+	Convey("given no audit event exists in the provided context", t, func() {
+		producer := &OutboundProducerMock{}
+
+		auditor := New(producer, namespace, "test")
+
+		// record the audit event
+		err := auditor.Record(context.Background(), auditAction, auditResult, nil)
+
+		expectedErr := auditError("no event found in context.Context", auditAction, nil)
```

Comment:
I can't help but feel that this test is too dependent on `auditError` working, and if it breaks (future change), then these tests will be testing the broken output against the broken output - and hence probably succeed.

Perhaps add a new `func Test_auditError(...)` before all these to test that `auditError` returns something sane/expected/non-"".

Reply:
Good idea. I have to say I am not a fan of Go approach errors - sometimes you want to pass back values that give the error some context but the only way to do that is building up a string which feels a bit naff. I guess I could create an audit error type which takes these values but ultimately it is string concatenation again.